lxd/acme/acme.go

```diff
@@ -0,0 +1,191 @@
+package acme
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/go-acme/lego/acme"
+	"github.com/go-acme/lego/certcrypto"
+	"github.com/go-acme/lego/certificate"
+	"github.com/go-acme/lego/lego"
+	"github.com/go-acme/lego/registration"
+
+	"github.com/lxc/lxd/lxd/state"
+	"github.com/lxc/lxd/lxd/util"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/logger"
+)
+
+// retries describes the number of retries after which LXD will give up registering a user or
+// issuing a new certificate. The number 5 was chosen because Let's Encrypt has a limit of 5
+// failures per account, per hostname, per hour.
+const retries = 5
+
+// ClusterCertFilename describes the filename of the new certificate which is stored in case it
+// cannot be distributed in a cluster due to offline members. LXD will try to distribute this
+// certificate at a later stage.
+const ClusterCertFilename = "cluster.crt.new"
+
+func certificateNeedsUpdate(domain string, cert *x509.Certificate) bool {
+	return shared.StringInSlice(domain, cert.DNSNames) && !time.Now().After(cert.NotAfter.Add(-time.Hour*24*30))
+}
+
+// UpdateCertificate updates the certificate.
+func UpdateCertificate(s *state.State, provider HTTP01Provider, clustered bool, domain string, email string, caURL string) (*certificate.Resource, error) {
+	clusterCertFilename := shared.VarPath(ClusterCertFilename)
+
+	l := logger.AddContext(logger.Log, logger.Ctx{"domain": domain, "caURL": caURL})
+
+	// If clusterCertFilename exists, it means that a previously issued certificate couldn't be
+	// distributed to all cluster members and was therefore kept back. In this case, don't issue
+	// a new certificate but return the previously issued one.
+	if clustered && shared.PathExists(clusterCertFilename) {
+		keyFilename := shared.VarPath("cluster.key")
+
+		clusterCert, err := os.ReadFile(clusterCertFilename)
+		if err != nil {
+			return nil, fmt.Errorf("Failed reading cluster certificate file: %w", err)
+		}
+
+		key, err := os.ReadFile(keyFilename)
+		if err != nil {
+			return nil, fmt.Errorf("Failed reading cluster key file: %w", err)
+		}
+
+		keyPair, err := tls.X509KeyPair(clusterCert, key)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to get keypair: %w", err)
+		}
+
+		cert, err := x509.ParseCertificate(keyPair.Certificate[0])
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse certificate: %w", err)
+		}
+
+		// If the domain matches the certificate's DNS names, and the certificate is valid for more
+		// than 30 days, return it.
+		if certificateNeedsUpdate(domain, cert) {
```

Comment:
This looks wrong. If the certificate was issued yesterday but wasn't able to be distributed, then it won't need to be updated. So this will mean we will then issue a new certificate rather than using the recently issued one.